File: stacklayoutmanager/src/main/java/com/maxxipoint/layoutmanager/adapter/StackAdapter.java
package com.maxxipoint.layoutmanager.adapter;

import android.text.Layout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import com.maxxipoint.layoutmanager.R;
import com.maxxipoint.layoutmanager.StackActivity;
import com.maxxipoint.layoutmanager.stack.CardConfig;
import com.maxxipoint.layoutmanager.stack.CardItemTouchHelperCallback;
import com.maxxipoint.layoutmanager.stack.CardLayoutManager;
import com.maxxipoint.layoutmanager.stack.OnSwipeListener;

import java.util.ArrayList;
import java.util.List;

public class StackAdapter extends RecyclerView.Adapter<StackAdapter.MyViewHolder> {

    private List<Integer> list = new ArrayList<>();
    private List<String> stackLists;

    public StackAdapter() {
//        this.stackLists = stackLists;
        list.add(R.drawable.shape_constellation);
        list.add(R.drawable.shape_age);
        list.add(R.drawable.shape_constellation);
        list.add(R.drawable.shape_age);
        list.add(R.drawable.shape_constellation);
        list.add(R.drawable.shape_age);
        list.add(R.drawable.shape_constellation);
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View inflate = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_stack, parent, false);
        return new MyViewHolder(inflate);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {

        holder.recyclerView.setAdapter(new MyAdapter(list));
        CardItemTouchHelperCallback cardCallback = new CardItemTouchHelperCallback(holder.recyclerView.getAdapter(), list);
        cardCallback.setOnSwipedListener(new OnSwipeListener<Integer>() {

            @Override
            public void onSwiping(RecyclerView.ViewHolder viewHolder, float ratio, int direction) {
//                StackActivity.MyAdapter.MyViewHolder myHolder = (StackActivity.MyAdapter.MyViewHolder) viewHolder;
//                viewHolder.itemView.setAlpha(1 - Math.abs(ratio) * 0.2f);
//                if (direction == CardConfig.SWIPING_LEFT) {
//                    myHolder.dislikeImageView.setAlpha(Math.abs(ratio));
//                } else if (direction == CardConfig.SWIPING_RIGHT) {
//                    myHolder.likeImageView.setAlpha(Math.abs(ratio));
//                } else {
//                    myHolder.dislikeImageView.setAlpha(0f);
//                    myHolder.likeImageView.setAlpha(0f);
//                }
            }

            @Override
            public void onSwiped(RecyclerView.ViewHolder viewHolder, Integer o, int direction) {
//                StackActivity.MyAdapter.MyViewHolder myHolder = (StackActivity.MyAdapter.MyViewHolder) viewHolder;
//                viewHolder.itemView.setAlpha(1f);
//                myHolder.dislikeImageView.setAlpha(0f);
//                myHolder.likeImageView.setAlpha(0f);
//                Toast.makeText(StackActivity.this, direction == CardConfig.SWIPED_LEFT ? "swiped left" : "swiped right", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onSwipedClear() { // 数据全部清空回调
//                Toast.makeText(StackActivity.this, "data clear", Toast.LENGTH_SHORT).show();
//                recyclerView.postDelayed(new Runnable() {
//                    @Override
//                    public void run() {
//                        initData();
//                        recyclerView.getAdapter().notifyDataSetChanged();
//                    }
//                }, 3000L);
            }
        });

        final ItemTouchHelper touchHelper = new ItemTouchHelper(cardCallback);
        final CardLayoutManager cardLayoutManager = new CardLayoutManager(holder.recyclerView, touchHelper);
        holder.recyclerView.setLayoutManager(cardLayoutManager);
        touchHelper.attachToRecyclerView(holder.recyclerView);
    }

    @Override
    public int getItemCount() {
        return 10;
    }

    class MyViewHolder extends RecyclerView.ViewHolder{

        RecyclerView recyclerView;

        public MyViewHolder(@NonNull View itemView) {
            super(itemView);
            recyclerView = itemView.findViewById(R.id.recycler_view);
            recyclerView.setItemAnimator(new DefaultItemAnimator());
        }
    }
}

File: stacklayoutmanager/src/main/java/com/maxxipoint/layoutmanager/adapter/StackAdapter.kt
package com.maxxipoint.layoutmanager.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.maxxipoint.layoutmanager.R
import com.maxxipoint.layoutmanager.bean.SlideBean
import com.maxxipoint.layoutmanager.use.stack.ItemTouchHelperCallback
import com.maxxipoint.layoutmanager.use.stack.OnStackListener
import com.maxxipoint.layoutmanager.use.stack.StackLayoutManager
import com.maxxipoint.layoutmanager.use.tool.TouchRecyclerView

class StackAdapter: RecyclerView.Adapter<RecyclerView.ViewHolder>(), OnStackListener {

    private val mList = mutableListOf<SlideBean?>()

    init {
        val bgs = intArrayOf(
            R.mipmap.img_slide_1,
            R.mipmap.img_slide_2,
            R.mipmap.img_slide_3,
            R.mipmap.img_slide_4,
            R.mipmap.img_slide_5,
            R.mipmap.img_slide_6
        )

        for (i in 0..5) {
            mList.add(SlideBean(bgs[i]))
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == 0) {
            val inflate = LayoutInflater.from(parent.context).inflate(R.layout.item_stack, parent, false)
            MyViewHolder(inflate)
        } else {
            val inflate = LayoutInflater.from(parent.context).inflate(R.layout.item_card, parent, false)
            TwoViewHolder(inflate)
        }
    }

    override fun getItemCount(): Int = 20

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is MyViewHolder) {
            holder.recyclerView.adapter = SlideChilAdapter(mList)
            holder.recyclerView.adapter?.let {
                val callback = ItemTouchHelperCallback<SlideBean?>(it, mList)
                val mItemTouchHelper = ItemTouchHelper(callback)
                val mSlideLayoutManager = StackLayoutManager(callback, holder.recyclerView, mItemTouchHelper)
                mSlideLayoutManager.setOnStackListener(this)
                mItemTouchHelper.attachToRecyclerView(holder.recyclerView)
                holder.recyclerView.layoutManager = mSlideLayoutManager
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 2) 0 else 1
    }

    override fun onSliding(viewHolder: RecyclerView.ViewHolder?, ratio: Float, direction: Int) {}

    override fun onSlided(viewHolder: RecyclerView.ViewHolder?, direction: Int) {}

    override fun onItemClick(position: Int) {

    }

    internal class MyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var recyclerView: TouchRecyclerView = itemView.findViewById(R.id.recycler_view)

        init {
            recyclerView.itemAnimator = DefaultItemAnimator()
        }
    }

    internal class TwoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}